scripts/rvpa.1
.Dd November 28, 2018
.Dt RVPA 1
.Os Linux
.Sh NAME
.Nm rvpa
.Nd analyze a trace file produced by a program compiled with
.Tn RV-Predict/C
instrumentation using
.Xr rvpc 1
.Sh SYNOPSIS
.Nm 
.Op Fl h
.Op Fl n
.Op Fl Fl help
.Op Fl Fl window Ar size
.Op Fl Fl html-dir Ar directory
.Op Fl Fl interrupts-target Ar number
.Op Fl Fl max-interrupt-depth Ar depth 
.Op Fl Fl no-shorten
.Op Fl Fl no-symbol
.Op Fl Fl no-system
.Op Fl Fl no-trim
.Op Fl Fl output=...
.Op Fl -
.Ar program
.Op Ar trace-file
.Sh DESCRIPTION
.Nm
analyzes a trace file left by
.Ar program Ap s
.Tn RV-Predict/C
instrumentation.
.Nm
reports any data races that it predicts on stderr.
.Pp
.Ar program
is used as a source for the DWARF debugging information
that
.Nm
needs to convert data & instruction addresses to names in its reports.
.Pp
If the optional argument
.Ar trace-file
is provided, then
.Nm
analyzes that trace file.
Otherwise,
.Nm
analyzes the trace file
.Pa rvpredict.trace 
left in the current directory by
.Ar program .
.Pp
Command-line options modify the behavior of
.Nm :
.Bl -tag -width "mmprompt-for-license"
.It Fl h | Fl Fl help
Print a help message and exit immediately.
.It Fl n
Print the Java command that would be run.
Perform no analysis.
.It Fl Fl html-dir Ar directory | Fl Fl html-dir Ns = Ns Ar directory
Reports will be written in HTML format under
.Ar directory .
.Pp
Each
.Fl Fl html-dir
parameter overrides all previous
.Fl Fl html-dir
and
.Fl Fl output
parameters on the command line.
.It Fl Fl max-interrupt-depth Ar depth | Fl Fl max-interrupt-depth Ns = Ns Ar depth
In each trace window, look for schedules where interrupts
.Dq stack
no more than
.Ar depth
deep, or no deeper than the trace events in that window actually show,
whichever is deeper.
A
.Ar depth
of 0 means no maximum.
A
.Ar depth
of 1 means an interruption of a thread, a
.Ar depth
of 2 means an interruption of an interruption of a thread, and so on.
.It Fl Fl interrupts-target Ar number | Fl Fl interrupt-target Ar number
.It Fl Fl interrupts-target Ns = Ns Ar number | Fl Fl interrupt-target Ns = Ns Ar number 
If there are fewer than
.Ar number
occurrences of any interrupt
.Va I
in a window
.Va W ,
then
.Tn RV-Predict/C
will copy occurrences of
.Va I
from prior windows into window
.Va W
until either there are
.Ar number
occurrences in
.Va W
or there are no more occurrences to copy.
.Pp
The default setting of
.Ar number
is 0:
no occurrences are copied from prior windows.
Increasing
.Ar number
can improve the prediction power of
.Tn RV-Predict/C ,
however, it will also prolong computation for each window.
.It Fl Fl window Ar size | Fl Fl window Ns = Ns Ar size
.Tn RV-Predict/C
segments
.Ar program Ap s
event trace into
.Dq windows
and analyzes one window at a time.
Windows are measured in the number of events they contain.
This option sets the number of events per window to
.Ar size ;
the default is 2000.
.Pp
There is a trade-off between prediction power and computation
time:
short windows take less time to analyze than long windows, however,
.Tn RV-Predict/C
may predict fewer data races using short windows than using long
windows on the same trace.
.It Fl Fl no-shorten
in reports,
long paths to source files are not shortened by replacing the first 
path components with an ellipsis,
.Dq ... .
.It Fl Fl no-symbol
disables symbolization altogether: instruction & data addresses
are not converted to file names, line and column numbers.
Signal numbers are not converted to symbols such as
.Dv SIGINT
and
.Dv SIGSEGV .
When
.Fl Fl no-symbol
is passed, options
.Fl Fl no-shorten ,
.Fl Fl no-system ,
and
.Fl Fl no-trim
have no effect.
.It Fl Fl no-system
stack frames in system libraries or header files are not elided. 
See
.Sx BUGS .
.It Fl Fl no-trim
stops
.Tn RV-Predict/C 
from eliding all stack frames under
.Fn main ,
all frames under a thread's
.Fn start_routine ,
and all frames under a signal handler.
.It Fl Fl output=console|csv|json|raw
indicates the desired report format.
.Dq console
is the default.
.Dq json
produces a report in JSON format that is suitable for
.Xr rv-html-report
to read.
.Dq csv
produces a table of comma-separated values, CSV, suitable
for import into spreadsheets and relational databases.
.Dq raw
produces unrefined JSON:
.Nm
performs no symbolization, no eliding of system frames, no trimming of
pathnames, regardless of the other settings on the command line.
.Pp
Each
.Fl Fl html-dir
parameter overrides all previous
.Fl Fl html-dir
and
.Fl Fl output
parameters on the command line.
.It Fl -
indicates the last option
for
.Nm .
Subsequent command-line words are treated
like the executable that
.Nm
should analyze.
.Fl -
is useful for analyzing an executable whose name
begins with
.Fl - .
.El
.\" This next command is for sections 1, 6, 7 and 8 only.
.Sh ENVIRONMENT
.Bl -tag -width "RVP_TRACE_FILE"
.It Ev RVP_TRACE_FILE
.Nm
ignores the setting of
.Ev RVP_TRACE_FILE
in the user's environment.
.It Ev RVP_ANALYSIS_ARGS
.Nm
treats the content of
.Ev RVP_ANALYSIS_ARGS
as if it appeared at the beginning of its command-line arguments.
Actual command-line arguments will override the arguments 
set by
.Ev RVP_ANALYSIS_ARGS .
.It Ev RVP_PROGRESS
If
.Ev RVP_PROGRESS 
is set to
.Dq yes ,
then
.Nm
prints messages about the progress that analysis is making.
Ordinarily the progress messages tell the next event window
that will be processed:
.Dl -- Window 1 --
.Dl -- Window 2 --
.Dl . . .
.It Ev RVP_WINDOW_SIZE
.Nm
treats this as the default window size.
.Ev RVP_WINDOW_SIZE takes precedence over window-size arguments in
.Ev RVP_ANALYSIS_ARGS ,
however, it may be overridden with a command-line argument.
.Op Fl Fl window Ar size
on the command line.
.It Ev RVP_REPORT_FILE
.Nm
writes its report to the file named by
.Ev RVP_REPORT_FILE ,
if the variable is set in the environment.
.Nm
quits with an error if both 
.Ev RVP_REPORT_FILE
and
.Ev RVP_REPORT_FD
are set in the environment.
.It Ev RVP_REPORT_FD
.Nm
writes its report to the file descriptor given by
.Ev RVP_REPORT_FD ,
if the variable is set in the environment.
.Nm
quits with an error if both 
.Ev RVP_REPORT_FILE
and
.Ev RVP_REPORT_FD
are set in the environment.
.El
.Sh FILES
.Bl -tag -width "/usr/share/examples/rv-predict-c/"
.It Pa rvpredict.trace
.Nm
looks in the directory where it is run for a trace file by this
name.
.It Pa /usr/share/doc/rv-predict-c/USERS-MANUAL.md
.Tn RV-Predict/C
Users Manual
.It Pa /usr/share/rv-predict-c/rv-predict.jar
the Java program implementing
.Tn RV-Predict/C
data-race prediction
.El
.Sh EXAMPLES
Build the demonstration program
called
.Nm lpcq
in
.Pa /usr/share/examples/rv-predict-c/c11
with
.Tn RV-Predict/C
instrumentation:
.Bd -literal
$ rvpc -o lpcq lpcq.c lpcq_main.c signals.c
.Ed
.Pp
Run it, and look for the trace file:
.Bd -literal
$ ./lpcq
read item 0
read item 1
read item 2
read item 3
read item 4
$ ls -l rvpredict.trace 
-rw------- 1 johndoe johndoe 6164 Jul 19 17:27 rvpredict.trace
.Ed
.Pp
Finally, analyze the trace:
.Bd -literal
$ rvpa ./lpcq
Data race on q.tailp at lpcq_main.c;main:
    Read in thread 2
      > in lpcq_get at .../c11/lpcq.c:26:19
        in consume at .../c11/lpcq_main.c:104
    Thread 2 created by thread 1
        in main at .../c11/lpcq_main.c:230

    Write in thread 1
      > in lpcq_put at .../c11/lpcq.c:48
        in produce at .../c11/lpcq_main.c:164
        in main at .../c11/lpcq_main.c:243
    Thread 1 is the main thread


Data race on [0x0000000000612080]:
    Read in thread 2
      > in lpcq_get at .../c11/lpcq.c:34:2
        in consume at .../c11/lpcq_main.c:104
    Thread 2 created by thread 1
        in main at .../c11/lpcq_main.c:230

    Write in thread 1
      > in lpcq_put at .../c11/lpcq.c:49
        in produce at .../c11/lpcq_main.c:164
        in main at .../c11/lpcq_main.c:243
    Thread 1 is the main thread
.Ed
.Pp
Several example programs are in
.Pa /usr/share/examples/rv-predict-c/ .
.Sh EXIT STATUS
If
.Nm
cannot parse its arguments, it returns 1.
If
.Nm
cannot find a
.Tn Java 
runtime, it returns 2.
.Nm
returns 3 if it finds
.Tn Java ,
but the version is not late enough.
.Sh SEE ALSO
.Xr rvpc 1
.Sh HISTORY
.Tn RV-Predict/C
1.9 was released in February 2018.
.Pp
.Tn RV-Predict/C
2.0 was released in July 2018.
.Sh AUTHORS
.An "Runtime Verification, Inc." Aq support@runtimeverification.com
.Sh BUGS
In this release, stack frames in system libraries and header files are
always symbolized, so the option
.Fl Fl no-system
has no effect.


debian/changelog
rv-predict-c (2.1.0-1) xenial; urgency=medium

  * Fix a bug where programs that `fork(2)` would hang.

  * Port to QNX.  Provide `qrvpc` for cross-building for QNX for x86_64.

  * Add an `ld(1)` replacement, `rvpld`.

  * Several improvements to the function and appearance of the HTML
    report.

  * Fix poor wording in reports that mention locks. 

  * Add synopses for the four categories of data race Predict detects.

  * Fix a bug that crashed some programs that used the POSIX
    thread-local storage (TLS) APIs (`pthread_getspecific(3)`,
    `pthread_setspecific(3)`).

  * Fix a but that crashed programs linked to dynamic libraries whose
    initializers used POSIX mutexes.

  * Separately instrument both the native Linux `signal(3)`, which is
    implemented using `sigaction(2)`, and the System V variant that is
    provided by `signal(2)`.

  * Support the `sigaction(2)` flag `SA_RESETHAND`.

  * Fix a bug that crashed some programs that used `memset(3)`.

  * Reduce report clutter: do not prepend `.../` to abbreviated paths.

  * Let the operator direct reports to a file or to a file descriptor
    using environment variables `RVP_REPORT_FILE` and `RVP_REPORT_FD`,
    respectively.

 -- Runtime Verification, Inc. <support@runtimeverification.com>  Thu, 13 Sep 2018 12:08:21 -0500

rv-predict-c (2.0.1-1) xenial; urgency=medium

  * Fix a defect in the report symbolization that broke some of
    the examples.

 -- Runtime Verification, Inc. <support@runtimeverification.com>  Wed, 01 Aug 2018 19:53:00 -0600

rv-predict-c (2.0-1) xenial; urgency=medium

  * Improve compatibility with LLVM coverage tools.

  * In data-race reports, print the source filename and line number,
    but don't print the column number, because LLVM appears to generate
    incorrect column numbers when coverage is enabled.  Do not print
    line number 0.

  * Avoid recording program traces where a program appears to enter a
    signal handler when the signal mask does not allow, or to read a
    signal mask that is inconsistent with the masks that were written.

  * Make `rvpsymbolize` correctly symbolize inputs that contain more
    than one function address.

  * Add to `rvpa(1)` an option `--max-interrupt-depth depth` that limits
    the level of recursion Predict will explore.

  * Print correct stack traces for interrupted threads and signals.  Fix
    a bug where a function entered in a previous window may not have
    appeared in a data race report for the current window.

  * Change the appearance of the HTML report and add controls to the
    data-race detail display for moving to the previous/next detailed
    display.

  * Introduce an interrupt personality, `tiaic`, for the Texas
    Instruments ARM Interrupt Controller.

  * Add an option, --interrupts-target=N (and variants), that tells how
    many times to try to run each interrupt in each window by copying
    interrupt traces from prior windows, if necessary.  Describe the
    option in the manual page.

  * Add {} to more instruction pointers in the symbol-friendly output of
    `rvpdump` so that `rvpsymbolize` will turn the pointers to function
    names.

  * Fix bugs that sometimes caused Predict-instrumented programs to hang
    when they received a signal such as SIGINT (Control-C).

  * Improve the examples for the Renesas 78k0 simulation.  Improve
    interrupt examples and their documentation.

  * Improve performance of Predict's data-race analysis.

  * Provide an example from Volker Stolz.

  * Add modular simulations for spontaneous signal entry.  Document the
    signal simulations.

  * Streamline the production of HTML reports.

  * Add options for adjusting the timeouts used by the data-race
    analysis.

  * Install `rv-html-report` as `rvp-html-report` to avoid conflicts
    with RV-Match.

  * Add instrumentation to the `signal(2)` system call.

  * Fix some bugs in the `sigaction(2)` instrumentation.

  * When the instrumented application is cancelled with, say, Control-C,
    don't cancel the analysis, too.

  * Cancel analysis if, after the application program has finished
    running, we receive another signal like SIGINT (Control-C), SIGHUP,
    SIGQUIT, et cetera.

  * Check the status of the analysis process and squeak on `stderr` if
    it exited on a signal.

  * Correctly forward `rvpc` command-line arguments that contain
    whitespace to `clang`.

  * Fix some bugs affecting atomic 8- and 16-bit stores and atomic
    16-bit read-modify-write.

  * Fix some data-race analysis crashes.

  * Add a `--version` command-line flag to `rvpc` and `rvpc++` commands.

  * Add a command-line option, `--parallel-smt #`, that tells how many
    threads to run the SMT solver in.

  * Improve the `rvpdump(1)` manual page.

  * Prevent a diagnostic assertion from occurring when the Predict
    runtime expected (incorrectly) for SIGKILL or SIGSTOP to be
    masked---they cannot be masked.

  * Perform online analysis by default.  That is, run the instrumented
    program and the data-race analysis concurrently.

 -- Runtime Verification, Inc. <support@runtimeverification.com>  Tue, 31 Jul 2018 23:22:00 -0600

rv-predict-c (1.9-1) xenial; urgency=medium

  * Miscellaneous bug fixes and new features.

 -- Runtime Verification, Inc. <support@runtimeverification.com>  Thu, 08 Feb 2018 16:58:07 -0600

rv-predict-c (1.8.40-1) xenial; urgency=medium

  * Miscellaneous bug fixes and new features.

 -- Runtime Verification, Inc. <support@runtimeverification.com>  Wed, 07 Feb 2018 12:44:19 -0600

rv-predict-c (1.8.39-1) xenial; urgency=medium

  * Miscellaneous bug fixes and new features.

 -- Runtime Verification, Inc. <support@runtimeverification.com>  Wed, 31 Jan 2018 21:19:43 -0600

rv-predict-c (1.8.38-1) xenial; urgency=medium

  * Miscellaneous bug fixes and new features.

 -- Runtime Verification, Inc. <support@runtimeverification.com>  Wed, 31 Jan 2018 18:13:57 -0600

rv-predict-c (1.8.37-1) xenial; urgency=medium


scripts/rvpa.sh
#!/bin/sh

set -e
set -u

only_print_command=no
html_dir=
analyze_passthrough=
symbolize_passthrough=
sharedir=$(dirname $0)/../share/rv-predict-c

_usage()
{
	cat 1>&2 <<EOF
usage: $(basename $0) [--window size] [--no-shorten|--no-symbol|--no-trim]
    [--html-dir directory] [--interrupts-target number]
    [--max-interrupt-depth number] [--]
    program [trace-file]
EOF
}

usage()
{
	_usage
	exit 1
}

help()
{
	_usage
	cat 1>&2 <<EOF

For more information, see the manual page, $(basename $0)(1).

EOF
	exit 1
}

rvpredict()
{
	pfx=

	if [ ${only_print_command:-no} = yes ]; then
		pfx=echo
		_java=java
	elif which java >/dev/null; then
		# found java executable in PATH
		_java=java
	elif [ "${JAVA_HOME:-}/bin/java" != "/bin/java" -a \
	       -x "${JAVA_HOME:-}/bin/java" ]; then
		# found java executable in JAVA_HOME
		_java="${JAVA_HOME:-}/bin/java"
	else
		cat 1>&2 <<EOF
RV Predict requires Java ${min_version} to run but Java was not detected.
Please either add it to PATH or set the JAVA_HOME environment variable.
EOF
		exit 2
	fi

	${pfx} ${_java} -ea -jar ${sharedir}/rv-predict.jar "$@"
}

do_output()
{
	[ ${RVP_REPORT_FILE:-none} != none ] \
	    && cat > ${RVP_REPORT_FILE} \
	    || cat 1>&${RVP_REPORT_FD:-2}
}

symbolize()
{
	if [ ${raw:-no} = yes ]; then
		cat
		return
	fi
	rvpsymbolize-json ${symbolize_passthrough} "$@" | \
	{ [ ${filter_trim:-yes} = yes ] && rvptrimframe || cat ; } | \
	{ [ ${filter_shorten:-yes} = yes ] && rvpshortenpaths || cat ; } | \
	RV_ISSUE_REPORT=/dev/stdout \
	    rvp-error ${sharedir}/${output_format:-console}-metadata.json | \
	{ [ x${html_dir:-} != x ] \
	    && rvp-html-report -o ${html_dir} /dev/stdin \
	    || do_output ; }
}

if [ ${RVP_REPORT_FD:-none} != none -a ${RVP_REPORT_FILE:-none} != none ]; then
	echo "$(basename $0): RVP_REPORT_FD conflicts with RVP_REPORT_FILE, set only one." 2>&1
	exit 1
fi

if [ ${RVP_REPORT_FD:-none} != none ]; then
	if [ -n "$(echo -n "$RVP_REPORT_FD" | sed 's/^[0-9]\+$//g')" ]; then
		echo "$(basename $0): malformed RVP_REPORT_FD: expected decimal digits, read '${RVP_REPORT_FD}'" 2>&1
		exit 1
	fi
fi

if [ ${RVP_PROGRESS:-no} = yes ]; then
	set -- "--progress" "$@"
fi

if [ ${RVP_WINDOW_SIZE:-none} != none ]; then
	if [ -n "$(echo -n "$RVP_WINDOW_SIZE" | sed 's/^[0-9]\+$//g')" ]; then
		echo "$(basename $0): malformed RVP_WINDOW_SIZE: expected decimal digits, read '${RVP_WINDOW_SIZE}'" 2>&1
		exit 1
	fi
	set -- "--window" ${RVP_WINDOW_SIZE} "$@"
fi

if [ -n "${RVP_ANALYSIS_ARGS:-}" ]; then
	set -- ${RVP_ANALYSIS_ARGS} "$@"
fi

if [ x${RVP_HTML_DIR:-} != x ]; then
	output_format=json
	html_dir=${RVP_HTML_DIR:-}
fi

while [ $# -ge 1 ]; do
	case $1 in
	-n)
		only_print_command=yes
		shift
		;;
	--no-symbol)
		symbolize_passthrough="${symbolize_passthrough:-} -S"
		shift
		;;
	--no-shorten|--no-trim)
		eval filter_${1##--no-}=no
		shift
		;;
	--html-dir)
		output_format=json
		shift
		html_dir="$1"
		shift
		;;
	--html-dir=*)
		output_format=json
		eval html_dir=${1##--html-dir=}
		shift
		;;
	--output=raw)
		raw=yes
		html_dir=
		shift
		;;
	--output=*)
		html_dir=
		eval output_format=${1##--output=}
		shift
		case $output_format in
		console|csv|json)
			;;
		*)
			echo "$(basename $0): unknown output format "
			    "'${output_format}'" 1>&2
			exit 1
			;;
		esac
		;;
	--interrupt-target=*|--interrupts-target=*)
		interrupts_target=${1##--interrupt-target=}
		interrupts_target=${interrupts_target##--interrupts-target=}
		shift
		;;
	--interrupt-target|--interrupts-target)
		shift
		interrupts_target="--desired-interrupts-per-signal-and-window $1"
		shift
		;;
	--parallel-smt)
		shift
		parallel="--parallel-smt $1"
		shift
		;;
	--window=*)
		window="--window ${1##--window=}"
		shift
		;;
	--window)
		shift
		window="--window $1"
		shift
		;;
	--global-timeout|--solver-timeout|--window-timeout)
		analyze_passthrough="${analyze_passthrough:-} $1 $2"
		shift
		shift
		;;
	--max-interrupt-depth=*)
		analyze_passthrough="${analyze_passthrough:-} --max-interrupt-depth ${1##--max-interrupt-depth=}"
		shift
		;;
	--max-interrupt-depth)
		analyze_passthrough="${analyze_passthrough:-} $1 $2"
		shift
		shift
		;;
	-h|--help)
		shift
		help
		;;
	--debug|--progress)
		analyze_passthrough="${analyze_passthrough:-} $1"
		shift
		;;
	--)
		shift
		break
		;;
	*)	break
		;;
	esac
done

if [ $# -gt 2 ]; then
	echo "$(basename $0): too many arguments" 1>&2
	usage
elif [ $# -lt 1 ]; then
	echo "$(basename $0): too few arguments" 1>&2
	usage
fi

if [ $# -eq 2 ]; then
	trace_file=$2
fi

progname=$1
if [ ${progname##/} != ${progname} ]; then
	progpath=${progname}
else
	progpath=$(pwd)/${progname}
fi

if ! test -e ${progpath}; then
	echo "$1: File not found" 1>&2
	exit 1
fi

rvpredict ${analyze_passthrough:-} ${window:---window 2000} ${parallel:-} \
    ${interrupts_target:-} \
    --json-report \
    --compact-trace ${trace_file:-./rvpredict.trace} | \
    { [ ${only_print_command:-no} = no ] && symbolize $progpath 2>&1 || cat ; }
